src/pages/Graph/SummaryPanelNode.tsx

```diff
@@ -108,16 +109,40 @@ export default class SummaryPanelNode extends React.Component<SummaryPanelPropTy
     }
 
     const filters = ['request_count', 'request_error_count', 'tcp_sent', 'tcp_received'];
-    // For special service dest nodes we want to narrow the data to only TS with 'unknown' workloads (see the related
-    // comparator in getNodeDatapoints).
-    let byLabelsIn = this.isSpecialServiceDest(nodeMetricType) ? ['destination_workload'] : undefined;
-    let byLabelsOut = data.isRoot ? ['destination_service_namespace'] : undefined;
-
-    const promise = getNodeMetrics(nodeMetricType, target, props, filters, undefined, byLabelsIn, byLabelsOut);
-    this.metricsPromise = makeCancelablePromise(promise);
+    let promiseOut: Promise<Response<Metrics>>, promiseIn: Promise<Response<Metrics>>;
+    // set outgoing unless it is a non-root outsider (because they have no outgoing edges) or a
+    // service node (because they don't have "real" outgoing edges).
+    if (data.nodeType !== NodeType.SERVICE && (data.isRoot || !data.isOutsider)) {
+      // use source metrics for outgoing, except for:
+      // - unknown nodes (no source telemetry)
+      // - istio namespace nodes (no source telemetry)
+      const reporter: Reporter =
+        data.nodeType === NodeType.UNKNOWN || data.namespace === serverConfig().istioNamespace
+          ? 'destination'
+          : 'source';
+      const byLabels = data.isRoot ? ['destination_service_namespace'] : undefined;
+      promiseOut = getNodeMetrics(nodeMetricType, target, props, filters, 'outbound', reporter, undefined, byLabels);
+    } else {
+      promiseOut = Promise.resolve({ data: { metrics: {}, histograms: {} } });
+    }
+    // set incoming unless it is a root (because they have no incoming edges)
+    if (!data.isRoot) {
+      // use dest metrics for incoming, except for service nodes which need source metrics to capture source errors
+      const reporter: Reporter =
+        data.nodeType === NodeType.SERVICE && data.namespace !== serverConfig().istioNamespace
+          ? 'destination'
+          : 'source';
+      // For special service dest nodes we want to narrow the data to only TS with 'unknown' workloads (see the related
+      // comparator in getNodeDatapoints).
+      const byLabels = this.isSpecialServiceDest(nodeMetricType) ? ['destination_workload'] : undefined;
+      promiseIn = getNodeMetrics(nodeMetricType, target, props, filters, 'inbound', reporter, undefined, byLabels);
```

Comment:
Not sure, but perhaps you are missing the TCP case here.

Unless I'm misunderstanding, this new code is not completely equivalent to the old one. Reading the old code, I understand that here if `reporter` is resolved to `destination`, you will need an additional call to `getNodeMetrics` to correctly fetch TCP data (since the old code is always using reporter=source for TCP).

Reply:
Yeah, looks like I missed a case

Reply:
Fixed in last commit, splitting TCP and HTTP in that case